engine/src/cmd/upgradeable_unit.cpp

```diff
@@ -0,0 +1,273 @@
+    if(upgrades.size() == 0) {
+        return std::vector<std::string>();
+    }
+
+    std::string upg_string = std::string(upgrades);
+
+    // Expected upgrades string is in the format of {<upgrade>;<mount offset>;<subunit offset}{...}
+    // Trim leading and trailing spaces
+    boost::algorithm::trim(upg_string);
+
+    // Remove all spaces and {
+    boost::replace_all(upg_string, " ", "");
+    boost::replace_all(upg_string, "{", "");
+
+    // Remove last }
+    boost::replace_last(upg_string, "}", "");
+
+    const std::string delimiter = "}";
+
+    std::vector<std::string> upgrades_vector;
+    boost::split(upgrades_vector, upg_string, boost::is_any_of(delimiter));
+
+    return upgrades_vector;
+}
+
+// TODO: why do we have to use such kludges?!
+unsigned int convert_to_int(std::string s) {
+    if(s.size() == 0) {
+        return 0;
+    }
+
+    try {
+        return std::stoi(s);
+    } catch(...) {
+        return 0;
+    }
+}
+
+UpgradeableUnit::UpgradeableUnit()
+{
+
+}
+
+extern int GetModeFromName(const char *input_buffer);
+
+// TODO: remove unit parameter
+void UpgradeableUnit::UpgradeUnit(const std::string &upgrades) {
+    const std::string delimiter = ";";
+
+    std::vector<std::string> upgrades_vector = ParseUnitUpgrades(upgrades);
+
+    for(const std::string& upgrade : upgrades_vector) {
+        std::vector<std::string> upgrade_vector;
+        boost::split(upgrade_vector, upgrade, boost::is_any_of(delimiter));
+
+        const unsigned int mount_offset = convert_to_int(upgrade_vector[1]);
+        const unsigned int subunit_offset = convert_to_int(upgrade_vector[2]);
+
+        const Unit *upgradee = UnitConstCache::getCachedConst(StringIntKey(upgrade, FactionUtil::GetUpgradeFaction()));
+        if (!upgradee) {
+            upgradee = UnitConstCache::setCachedConst(StringIntKey(upgrade, FactionUtil::GetUpgradeFaction()),
+                    new Unit(upgrade.c_str(),
+                            true,
+                            FactionUtil::GetUpgradeFaction()));
+        }
+
+        double percent = 1.0;
+        const int mode = GetModeFromName(upgrade.c_str());
+
+        // TODO: change this when we make this a sub-class of unit
+        Unit *unit = static_cast<Unit *>(this);
+        unit->Upgrade(upgradee, mount_offset, subunit_offset, mode, true, percent, nullptr);
+    }
+}
+
+const Unit *getUnitFromUpgradeName(const string &upgradeName, int myUnitFaction = 0);
+
+static const std::string LOAD_FAILED = "LOAD_FAILED";
+
+
+bool UpgradeableUnit::UpgradeMounts(const Unit *up,
+                                    int mountoffset,
+                                    bool touchme,
+                                    bool downgrade,
+                                    int &numave,        // number of used parts?
+                                    double &percentage) // percentage of used parts
+{
+    Unit* unit = static_cast<Unit *>(this);
```

Comment:
I would prefer to see a dynamic pointer cast here, rather than a `static_cast`. You should be able to use `vega_dynamic_cast_ptr` from "vega_cast_utils.h".

Reply:
@BenjamenMeyer what do you think
Why use our own cast when we can use a well known, part of the language, one? 
Also, I've used a ton of static casts already in other parts of unit classes. If we make a decision here, I'll open an issue to change them all.

Reply:
You can use the built-in dynamic pointer cast operator if you want. The `vega_...` one is just designed to be a convenience wrapper for the built-in one. It logs a message and aborts the program if the cast attempt fails.

Reply:
Much better, thanks